Structural Changes for the YouTube Demo

AngelAlvie opened the pull request:
I have added a number of changes that should improve the structure and quality of the JavaScript code. 

Here are the changes I made:
1. Reorganized directory structure.
2. Changed the demo to load from various CDNs for dependencies, it would be faster for most users.
3. Upgraded JQuery and D3
4. Added separation of player and graph from the index page. This allows for us to reuse and modify these components for future use, or exchange the underlying libraries/API that they access without needing to change the rest of the code.
5. Helpful annotation of all the functions and descriptions of what they do inside JSDoc strings.
6. Updated to ES6 syntax for variables, functions, etc.
7. Reduced uses of `this` keyword. It is typically considered unreliable, as it depends on the caller's environment, not the initialized environment.
8. Updated state model: the demo is now broken down into 4 distinct states, `Loading`, `Searching`, `Recording`, `Playback`.
9. Utilization of new ES6 features such as promises to make concurrent tasks run faster.
10. Refactoring methods and their organization to improve readability, and reusability.
11. Fixed several cases of `callback hell` by flattening code wherever possible.

Here is a list of the following behavioral changes that an end user should be able to see as a result of the changes above:

1. Scripts should not block the loading of the page HTML, and should load faster since they are fetching their code asynchronously, and from a CDN.
2. Detector, YouTube Player and example videos should load asynchronously, thus the search component of the page should load faster.
3. Cursor saves the playing state of the player, so that when the user drags while the video is playing, the video pauses until they are finished dragging.
4. The example videos load together, so the videos that load are squished toward the front of the selected videos element.
5. The bug where zeroed data was being filled in the graph has been fixed, when the detector reports no data, no data is filled for that time, The curve just interpolates to the next point whenever possible.

AngelAlvie commented:
If you would like to see these changes in deployment you can check out: https://angelalvie.github.io/youtube-demo/

That is the gh-pages site corresponding to these changes.

AngelAlvie commented:
Okay, I realized that there was a side effect of the changes that I made that I didn't realize would break things. I had separated out the handlers for clicking a video and clicking the `start` button. As a result, the direct URL search feature is broken. This can be quicly fixed by adding the following line to the `startButtonClicked` method after we acquire the video ID from the URL:
`transitionToRecording(video_id);`

After the change, the `startButtonClicked` method should look like this:
```JavaScript
  /** Perform search after start button clicked. */
  const startButtonClicked = (_) => {
    $(".demo-message").hide();
    let video_id;
    if (state === self.States.SEARCHING) {
      
      const blob = document.getElementById("start-form").value;
      
      if (blob === "" || blob.includes("http://") || blob.includes("https://")) { // treat as URL

        console.log("registered");
        video_id = blob.split("v=")[1] || "";
        const ampersandPosition = video_id.indexOf("&");
        if (ampersandPosition !== -1) {
          video_id = video_id.substring(0, ampersandPosition);
        }
        transitionToRecording(video_id);

      } else { // treat as search
        const url = "https://www.googleapis.com/youtube/v3/search?part=snippet&type=video&key=" + API_KEY + "&maxResults=10&safeSearch=strict&q=" + blob;
        httpGetAsync(url)
          .then(addToSearchResults)
          .catch(ignore);
      }
    }
  };
```
However, since I updated the API key, the classical search feature now works (this is not a permanent fix though).

AngelAlvie commented:
While I was fixing this, I found that I wasn't getting a bad-url message when I inserted a bad URL. This is because in player.js, `onError` is misclassified as `onPlayerError`, and on the index side, in `transitionToRecording`, in the callback to player("play"), the part that handles an 'error' should use `showMessage("msg-bad-url")` instead of `showMessage("msg-player-failure")`.

AngelAlvie commented:
I just commit the fixes to my development branch that should fix the YouTube URL bug, Travis will push that new build to this branch momentarily

AngelAlvie commented:
It looks like there was an issue with travis overwriting the file, because the resolving ghpages commit was not added to the development branch, I will fix this now.

AngelAlvie commented:
Alright, on my production view [https://angelalvie.github.io/youtube-demo](https://angelalvie.github.io/youtube-demo), inputting a URL should work as it did before.

AngelAlvie commented:
I have fixed the invalid data graph bug, the graph now shows no data if there is none for that particular time period instead of interpolating it, or showing zeroes.

AngelAlvie commented:
There is a stray console line, this next change should delete it.

AngelAlvie commented:
I am correcting a typo in the html. Where it says "this may take few minutes...", there was no 'a'. It should now say 'this may take a few minutes...'

AngelAlvie commented:
We found that the `onInitializeFailure` callback is called once before a success on firefox. I suppressed the `onInitializeFailure` promise rejection. This might be an issue with the javascript SDK on firefox.

j4y commented on .gitignore:
```diff
@@ -1,2 +1,9 @@
 .*
 !/.gitignore
+!/.travis.yml
+<<<<<<< HEAD
```
Looks like there is a merge conflict here in the .gitignore that needs to be cleared out.

AngelAlvie commented:
I have just pushed a change that will omit the browser check, and also fix the timing of the no face alert.

AngelAlvie commented:
This is so that craig can test with other browsers on https://angelalvie.github.io/youtube-demo/, originally the browser check is there to enforce that you are using the demo on chrome, firefox and opera.

AngelAlvie commented:
Last commit was to add the Browser check back to the deployed gh-pages, and add a new check for Edge, because the demo works on that browser.

AngelAlvie commented:
This newest commit adds a gray box over the graph where there was no data. This should make the graph a little more pleasing, there are some considerations though to look at for this change. For instance, how dark should the gray box be? and should we provide an explanation of what the box means?

umangmehta12 commented on README.md:
```diff
@@ -44,6 +44,59 @@ $ python -m http.server 8000
 
+## Components
+
+This demo is broken down into three components under `public/javascripts/`.
```
I don't see a directory called public is that intended? Are you planning to creating a public directory?

AngelAlvie replied:
Good catch, the public directory is part of my own development branch. This should be just `javascripts/`

umangmehta12 commented on javascripts/index.js:
```diff
@@ -0,0 +1,575 @@
+      detector.addEventListener("onInitializeSuccess", () => {
+        resolve();
+      });
+      //detector.addEventListener("onInitializeFailure", () => {
```
Can we remove this commented code?

AngelAlvie replied:
This code is commented because it appears that there is a potential problem on the SDK side involving the `onInitializeFailure` callback on firefox. I was investigating this issue, and didn't want to make any assumptions about when it should be called. It appears firefox will call the `onInitializeFailure` callback once, before then firing the `onInitializeSuccess` callback. I had assumed that the SDK would only fire the `onInitalizeFailure` callback if there was a irrecoverable error from the SDK. I can remove it for now until the issue is either resolved, or the behavior of the callback is actually correct.

umangmehta12 commented on javascripts/index.js:
```diff
@@ -0,0 +1,575 @@
+
+  /** Set ordering of initial videos to be in the same order as the video ids list. */
+  const sortVideos = () => {
+    // I'm not worried about efficency, since I know  the number of elements is bounded by a really small number (6).
```
Can the comment(if needed) be in third person?

umangmehta12 commented on javascripts/index.js:
```diff
@@ -0,0 +1,575 @@
+        </tr>
+      </table>`;    
+      $("#search-results").append(result);
+      $("#"+id).click({id: id}, onVideoClick);  // These are buttons that are being used in start button click. i don't know why there is one big function
```
Can we get rid of this comment?

umangmehta12 commented on javascripts/index.js:
```diff
@@ -0,0 +1,575 @@
+  const dragStartHandler = () => {
+    if (player("getPlayingState")) {
+      clearInterval(cursor_interval);
+      // I want to stop the player until we stop dragging, but keep the state of the player as we drag.
```
Can we change the comment (if needed) to third person?

umangmehta12 commented on javascripts/player.js:
```diff
@@ -0,0 +1,121 @@
+/** Here I want to define a Asynchronous Player object that supports a number of signals that can be sent to it
```
Can we change this comment to third person?

umangmehta12 commented on javascripts/player.js:
```diff
@@ -0,0 +1,121 @@
+/** Here I want to define a Asynchronous Player object that supports a number of signals that can be sent to it
+ */
+function AsyncPlayer() {
+  // This represents a closure, where I can define all of my variables that I need.
```
this one as well.

umangmehta12 commented on javascripts/player.js:
```diff
@@ -0,0 +1,121 @@
+      cb("network fail", null);
+      player.stopVideo();
+    }
+    // make cursor less buggy while video is paused
```
good to remove this comment, it indicates their is a bug?

umangmehta12 commented on javascripts/index.js:
```diff
@@ -0,0 +1,575 @@
+  let detector = null;
+
+  // The YouTube Data API KEY should really not be in the client, but oh well.
+  let API_KEY = "AIzaSyBw81iUUXQpYRuxSVmMc2jNkjv1tJwqHjc";
```
which account did we generate this API key from,can we can hide it?

umangmehta12 commented on javascripts/index.js:
```diff
@@ -0,0 +1,575 @@
+  /** Transition the page into the RECORDING state. 
+   * @param {string} video_id - internal youtube id for the video that we want to transition to playing */
+  const transitionToRecording = (video_id) => {
+    // Remove any demo messages we recieved
```
typo

j4y commented:
It looks like each deployment added a commit.  Is that intentional?  It looks like they will be included with this pull request.

AngelAlvie commented:
Yes, that is intentional, I have a development branch that I make changes to, and that in turn makes changes my gh-pages branch. The difference between the two branches is that the development branch contains some extra tools.

jmr76 commented:
Nice job. I like the code encapsulation. My only question is: how does this operate on older browsers, and do we care? Some older browsers don't support the ES6 syntax (let, const, -> {}, etc....)

AngelAlvie commented:
The unfortunate issue with the API security, is that if we want to have the code be on GitHub, we need to check the process of getting the search results into GitHub.

AngelAlvie commented:
Jay had told me that since this was a demo, we shouldn't need to worry much about legacy support. If this really becomes a concern I could probably trans-pile it using TypeScript to an older JS version, and add a polyfill for promises.

AngelAlvie commented:
Alright, I should have just now fixed the grammatical issues with the comments, moved files to `js/` and `css/` and added a documented solution for the API key issue.

AngelAlvie commented:
Alright, the changes that Umang had requested have been implemented.
